File: java/src/main/java/vsdk/toolkit/galerkin/processing/visitors/PowerAccumulatorVisitor.java
package vsdk.toolkit.galerkin.processing.visitors;

import vsdk.toolkit.common.ColorRgb;
import vsdk.toolkit.common.linealAlgebra.Numeric;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.galerkin.GalerkinElement;
import vsdk.toolkit.galerkin.GalerkinIterationMethod;
import vsdk.toolkit.galerkin.GalerkinState;

public final class PowerAccumulatorVisitor implements ClusterLeafVisitor {
    private ColorRgb sourceRadiance;
    private Vector3D samplePoint;
    private ColorRgb accumulatedRadiance;

    public PowerAccumulatorVisitor(ColorRgb inSourceRadiance, Vector3D inSamplePoint) {
        sourceRadiance = inSourceRadiance;
        samplePoint = inSamplePoint;
        accumulatedRadiance = new ColorRgb();
        accumulatedRadiance.clear();
    }

    public ColorRgb getAccumulatedRadiance() {
        return accumulatedRadiance;
    }

    /**
Accumulates this leaf contribution towards the visitor sample point.
The contribution is weighted by projected area (cosine term and patch area),
uses the radiance channel that matches the current Galerkin iteration mode,
and ignores intra-cluster visibility.
*/
    @Override
    public void visit(
        GalerkinElement galerkinElement,
        GalerkinState galerkinState)
    {
        Vector3D dir = new Vector3D();
        dir.subtraction(samplePoint, galerkinElement.patch.midPoint);
        float dist = dir.norm();
        float srcOs;
        if ( dist < Numeric.EPSILON ) {
            srcOs = 1.0f;
        }
        else {
            srcOs = dir.dotProduct(galerkinElement.patch.normal) / dist;
        }
        if ( srcOs <= 0.0f ) {
            return;
        }

        ColorRgb rad;
        if ( galerkinState.galerkinIterationMethod == GalerkinIterationMethod.GAUSS_SEIDEL ||
             galerkinState.galerkinIterationMethod == GalerkinIterationMethod.JACOBI ) {
            rad = galerkinElement.radiance[0];
        }
        else {
            rad = galerkinElement.unShotRadiance[0];
        }

        accumulatedRadiance.addScaled(sourceRadiance, srcOs * galerkinElement.area, rad);
    }
}

File: typescript/base/src/main/vsdk/toolkit/galerkin/processing/visitors/PowerAccumulatorVisitor.ts
import { ColorRgb } from "../../../common/color/ColorRgb";
import { Numeric } from "../../../common/linealAlgebra/Numeric";
import { Vector3D } from "../../../common/linealAlgebra/Vector3D";
import { GalerkinElement } from "../../GalerkinElement";
import { GalerkinIterationMethod } from "../../GalerkinIterationMethod";
import { GalerkinState } from "../../GalerkinState";
import { ClusterLeafVisitor } from "./ClusterLeafVisitor";

export class PowerAccumulatorVisitor implements ClusterLeafVisitor {
  private readonly sourceRadiance: ColorRgb;
  private readonly samplePoint: Vector3D;
  private readonly accumulatedRadiance: ColorRgb;

  public constructor(inSourceRadiance: ColorRgb, inSamplePoint: Vector3D) {
    this.sourceRadiance = inSourceRadiance;
    this.samplePoint = inSamplePoint;
    this.accumulatedRadiance = new ColorRgb();
    this.accumulatedRadiance.clear();
  }

  public getAccumulatedRadiance(): ColorRgb {
    return this.accumulatedRadiance;
  }

  public visit(
    galerkinElement: GalerkinElement,
    galerkinState: GalerkinState,
  ): void {
    const dir = new Vector3D();
    dir.subtraction(this.samplePoint, (galerkinElement.patch as NonNullable<GalerkinElement["patch"]>).midPoint);
    const dist = dir.norm();
    let srcOs: number;
    if (dist < Numeric.EPSILON) {
      srcOs = 1.0;
    }
    else {
      srcOs = dir.dotProduct((galerkinElement.patch as NonNullable<GalerkinElement["patch"]>).normal) / dist;
    }
    if (srcOs <= 0.0) {
      return;
    }

    let rad: ColorRgb;
    if (galerkinState.galerkinIterationMethod === GalerkinIterationMethod.GAUSS_SEIDEL
      || galerkinState.galerkinIterationMethod === GalerkinIterationMethod.JACOBI) {
      rad = (galerkinElement.radiance as ColorRgb[])[0]!;
    }
    else {
      rad = (galerkinElement.unShotRadiance as ColorRgb[])[0]!;
    }

    this.accumulatedRadiance.addScaled(this.sourceRadiance, srcOs * galerkinElement.area, rad);
  }
}
